grafast/dataplan-pg/src/steps/withPgClient.ts

```diff
@@ -3,17 +3,19 @@ import { constant, ExecutableStep } from "grafast";
 
 import type { PgClient, PgExecutor, WithPgClient } from "../executor";
 
-export type WithPgClientStepCallback<TData, TResult> = (
-  client: PgClient,
-  data: TData,
-) => Promise<TResult>;
+export type WithPgClientStepCallback<
+  TPgClient extends PgClient,
+  TData,
+  TResult,
```

Comment:
Here's roughly [the status quo on TypeScript Playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBACg5gUQB4QMYFdgHsBOUC8UA3hAFxTA7oQC+A3AFCiRTJqYCGARgDYQDKwCGAA8AFQB8BYgGchYcmJoMAlgDshOAGYdU0eAGEeKiBuIMoUAI7UcIABQcccGeSIUISYOTk51cKBoASnIYHCwAWxUZCBEONRAJBmUGTzBcYApwaAB1FWAAC0NjU2BBYQMOHh4uXQBrcQARDmAOABooMQAlCBl0HmApQnsLKFQSjVC4IxMNNtGAExaORWbW+aCCKTDI6Nju3v7BxjSMqC10NVRgFSw1KAB3fKLpieAm5Y6DvoGJEctPOxsDgpmwMMD5pYACRLVqkUaWAA+rBQ4O4fHKojEaw4SUsSKg9mxyygniEagWMigan6PCgyMuCwgWnUEAWUAA-NTafSoIzmaz2eQ1BAAG4QHBBSFjKo1erkPKFYqzMrySrVWqoBrE9adHo-QYbchgzi8ATycT6o5SIijHAQYDoHD3GnVKAcKnxECMFKoO5yUmozC4aSuuke90Jd1U+Am4GMP1qANQlYooHo83CES+fxDWTycgAIiEckL9AYiYDWmkTyVrxV9kB4NwHRTHQ9ICuhPGKvbm3wUg4QUYAHoR-ioAA9DkMIA). Note `TData` and `TResult` can be inferred from usage.

Adding a new generic at the beginning seems like it's a breaking change for no specific reason; e.g. if you call `withPgClient<SuperPgClient>(...)` [you'll get the error](https://www.typescriptlang.org/play/?#code/C4TwDgpgBACg5gUQB4QMYFdgHsBOUC8UA3hAFxTA7oQC+A3AFCiRTJqYCGARgDYQDKwCGAA8AFQB8BYgGchYcmJoMAlgDshOAGYdU0eAGEeKiBuIMoUAI7UcIABQcccGeSIUISYOTk51cKBoASnIYHCwAWxUZCBEONRAJBmVVDQhtXWh+dEgcQ2NTYChPITUAExlYOCMTMyILSygy+L17EKg1dAiudMYUzzBcIuZoAHUVYAALfNrgQWEDDh4eLl0Aa3EZwuKvUwqqmsKAGigxABEOYA4TsQAlCBl0HmApQnsG1AKNRS2NI4bmldFBcrv8ggQpGFItFYncHk8XowBkMoFp0GpUMAVFg1FAAO4TabVL7ATbE2Y7Ur7X7AG4g66ne6PZ4Sd6WTzsbA4UKIFAYLn-SwAEkBHFIDUsAB9WHzOLwBPJxPSko0oNL7OdLhxKXtKp1lmqoOiyhAtOoIGUoAB+DpPHiG42m82W8hqCAAN3SQUFUFQSxW63I4ymNPmYEWy1WqA2YhpdK1NyZCIkYPIbH53D4YfESZZ5ksOAgwHQOFx+vtHEq8RAfQYDFQOLkO05uGk5aglY7CQ7lXg6cwuEYDbUTaFYplnMzCuEIl8-lesnk5AAREI5Mv6PXG0UtNICSHyYURNlcjTWRz+bgTmOTpWQBioPZPrNb+D8FJ3Bw8RwJr6SQA6Gx0gcIghC8cgOGCOgoELYtS0sSCgkYAB6ZDVQAPStBggA) that `withPgClient` requires 3 generics. So if we're going to add this I think we should do so via the last position:

```suggestion
export type WithPgClientStepCallback<
  TData,
  TResult,
  TPgClient extends PgClient = PgClient,
```

Unless there's a good reason not to?

Further I think we should do a ridiculous TypeScript hack for typing and insert a function that does (essentially) nothing:

```ts
const $result = withCustomPgClient<SuperPgClient>()(
  executor,
  $data,
  async (pgClient, data) => {
    //   ^? SuperPgClient
  }
);
```

This allows you to pass the type for the client whilst still having the other two generics inferred.

https://www.typescriptlang.org/play/?#code/C4TwDgpgBACg5gUQB4QMYFdgHsBOUC8UA3lBAFxTA7rQC+A3AFCiRTJqYCGARgDYQBlYBDAAeACoA+AsSgBnYWArioDRgEsAdsJwAzTqmjwAwr3URtxRlCgBHGjhAAKTjjhyKJYUmAUFOLThVAEoKGBwsAFt1OQhRTk0QSSZaRg1tCD0DaAF0SBwTMwtgUh8LABM5WDhTc0siayhyhMMnUKhNdEjuTJS0iCQwXBKWaAB1dWAAC0K64CERY05eXm4DAGtRRvEAEU5gTgAabYAlCDl0XmBjm3FZ4tLhTUrq2uLj6UInVCLtZXvtIcmvtOMo9gdggRpOEojE4uIzhcrslGAMhjgSrp0JpUMB1FhNFAAO6TGY1X7ACQAkoDJ4valA3YgxmIy7ASRORoDDjYHBhRAoDC8m5QAAkzQOZEaNgAPmxBVw+IJFBJwZxJNKoHKnEyDo8KlVOistVBseUILotBBylAAPwdS68E1mi1Wm0UTQQABumWCItQy1WGwoE2m1IWYCWKzWqE2uqOUAR5zZjOpkmO7XYQp4-AjElZyKsNhwEGA6BwhKNTs4VQSID6jFQBIUpR5uBkVagNa7iS7VXgWcwuCYTc0LdFoPlPJzypEon8gU+sgUIgoACJhAo16oR83MTISWHyXMnNyhbggROgTWQDioN8KdfIfhpA0bD85kwbAB6b9QAB6tqNJwRKcJMUAfsUAB09iZM4Xi0hQnAhF+UAlmWFZdikwS7mOJQAGIyLoTC-jYZFkYB-SDMMUBYjieIEsSpLGOgChRNSVLHg8tIGq8FIcu0Wy3GqIpJki1yMByXIKry-KDsKjTiiCUrkXK8kznm8Yaqp97xvqzyGo6zrPK6no2vanZyi6lpmVAHrer6-qBjG6whqS4aKFGQaxqqzKJgW1yJmmGZQlO2ZKppAWvo06HloSh5km82gNqOLZBIQCUsWxkQcbk+Rpm0nI2GeQ44CKV7AXIt6oPekGAl2z7ReRdXAKhP5-pR5EgWBJQtTBDjwZQiFdihmqxZhnCobQGa4S2ADiMhwCRf7kRRQFAA


That said... It feels like it's the `executor` itself that should contain information on what the PgClient type is (since that's where the plan that gets it from `context()` resides). We might need TypeScript code generation for that to work reliably though.